Firestore/core/src/firebase/firestore/core/firestore_client.h

```diff
@@ -172,6 +172,13 @@ class FirestoreClient : public std::enable_shared_from_this<FirestoreClient> {
   }
   bool is_terminated() const;
 
+#if __cplusplus < 201402L
```

Comment:
Why not just have
```cpp
std::weak_ptr<FirestoreClient> weak_this = shared_from_this();
```
at the point of the call? While `weak_from_this` is the better solution in C++14, I'm not sure the benefits outweigh having this conditional logic.

Reply:
I was making it such that the caller could look like C++14 and we could just drop this block when C++14 became the floor. I'm happy to do it the other way (review usages of `shared_from_this` to see if any could be made `weak_from_this`).

Reply:
Not for this PR: perhaps we could have our own `enable_shared_from_this` for the time being. It would inherit from the standard library one and conditionally add `weak_from_this`. It's more work, but the benefit would be that the conditional logic is in just one place. No strong feelings about this, though.

Reply:
I considered that, but it seems worse than both iterations that are in this PR. Having a custom base class means that our classes no longer extend `std::enable_shared_from_this` which limits the self-documentary benefit.